roachpb/data.go

```diff
@@ -627,51 +627,54 @@ func MakePriority(userPriority float64) int32 {
 	//
-	// If pri > 1, let:
-	//   x = fraction of max such that choosing rand in [max*x, max)
-	//       is pri times more likely to be greater than random [0, max).
 	//
-	// For every random trial, if the normal priority is chosen < max*x,
-	// the user priority wins 100% of the time; otherwise, wins 50% of
-	// the time. Therefore, x fraction of the time, win is user(100%)
-	// and (1-x) fraction of the time win is user(50%).
+	// We use the exponential distribution. This distribution has the probability density function
+	//   PDF_lambda(x) = lambda * exp(-lambda * x)
+	// and the cumulative distribution function (i.e. probability that a random value is smaller
+	// than x):
+	//   CDF_lambda(x) = Integral_0^x PDF_lambda(x) dx
+	//                 = 1 - exp(-lambda * x)
 	//
-	//   x + (1-x) * 1/2 = pri / (1 + pri)
-	//   x/2 = pri / (1 + pri) - 1/2
-	//   x = (pri - 1) / (1 + pri)
+	// Let's assume we generate x from the exponential distribution with the lambda rate set to
+	// l1 and we generate y from the distribution with the rate set to l2. The probability that x
+	// wins is:
+	//    P[ x > y ] = Integral_0^inf Integral_0^x PDF_l1(x) PDF_l2(y) dy dx
+	//               = Integral_0^inf PDF_l1(x) Integral_0^x PDF_l2(y) dy dx
+	//               = Integral_0^inf PDF_l1(x) CDF_l2(x) dx
+	//               = Integral_0^inf PDF_l1(x) (1 - exp(-l2 * x)) dx
+	//               = 1 - Integral_0^inf l1 * exp(-(l1+l2) * x) dx
+	//               = 1 - l1 / (l1 + l2) * Integral_0^inf PDF_(l1+l2)(x) dx
+	//               = 1 - l1 / (l1 + l2)
+	//               = l2 / (l1 + l2)
 	//
-	// Otherwise, if pri < 1, let:
-	//   x = fraction of max such that choosing rand in [0, max*x)
-	//       is pri times more likely to be greater than random [0, max).
+	// We want this probability to be p1 / (p1 + p2) which we can get by setting
+	//    l1 = 1 / p1
+	//    l2 = 1 / p2
+	// It's easy to verify that (1/p2) / (1/p1 + 1/p2) = p1 / (p2 + p1).
 	//
-	// For every random trial, if the normal priority is chosen >= max*x,
-	// the user priority loses 100% of the time; otherwise, wins 50% of
-	// the time. Therefore, (1-x) fraction of the time, loss is user(100%)
-	// and x fraction of the time win is user(50%).
+	// We can generate an exponentially distributed value using (rand.ExpFloat64() / lambda).
+	// In our case this works out to simply rand.ExpFloat64() * userPriority.
 	//
-	//   x * 1/2 = pri / (1 + pri)
-	//   x = 2*pri / (1 + pri) = (pri - 1)/(pri + 1) + 1
-	pri := float64(userPriority)
-	x := (pri - 1) / (pri + 1)
-	if userPriority >= 1 {
-		return math.MaxInt32 - rand.Int31n(int32(float64(math.MaxInt32)*(1-x)))
-	}
-	return rand.Int31n(int32(float64(math.MaxInt32) * (1 + x)))
+	// To convert to an integer, we scale things to accomodate a few (10) standard deviations
+	// for the maximum priority. Note that the probability function falls exponentially.
+	val := rand.ExpFloat64() * userPriority / (10 * MaxUserPriority) * math.MaxInt32
```

Comment:
doesn't look like numerics gets in your way here, otherwise some extra parentheses are always helpful.